### ISSUE TITLE
Evaluate performance impact empty-closure lambda groups

### ISSUE THREAD

**Issue by heckej** [heckej/proguard#4]

**Comment by heckej** [comment]
# Evaluating the impact of empty closure lambda merging on the application size of the ComposeCodeViewer

## Summary

The application to which the lambda merging is applied is the [ComposeCodeViewer](https://github.com/JetBrains/compose-jb/tree/master/examples/codeviewer). This application is likely to gain some size improvements by optimising lambda's, as it uses [Compose for desktop](https://github.com/JetBrains/compose-jb), which encourages the usage of Kotlin lambda's.

To calculate the code size, we unzip the jars of the application, and count the total size of all class files: **85881117 bytes**.

After optimisation, the total size of all unzipped class files is **84689881 bytes**. That is **98.613%** of the original size, and therefore an improvement of **1.387%**.

## Used configuration

The ProGuard configuration that was used to optimise the application is the following with the lambda merging optimisation enabled (hard-coded):

```lang=pro
-ignorewarnings
-dontwarn java.**

-verbose
-dontobfuscate
-dontoptimize
-dontshrink

-keep public class org.jetbrains.codeviewer.MainKt {
    public static void main(java.lang.String[]);
}
```

## Notes

* The lambda groups must call the super constructor of the lambda class with the correct arity, such that no casting exceptions show up, when the Kotlin runtime checks whether it can cast the lambda group to the intended interface (see issue #12).
* Inner lambda's must be merged before outer lambda's (see issue #1).

**Comment by heckej** [comment]
Performance evaluations are available at [github.com/heckej/thesis-performance-evaluation](https://github.com/heckej/thesis-performance-evaluation).
